explorer/interpreter/type_checker.cpp

```diff
@@ -2863,6 +2863,24 @@ auto TypeChecker::TypeCheckStmt(Nonnull<Statement*> s,
       CARBON_RETURN_IF_ERROR(TypeCheckStmt(&while_stmt.body(), impl_scope));
       return Success();
     }
+    case StatementKind::For: {
+      auto& for_stmt = cast<For>(*s);
+      ImplScope inner_impl_scope;
+      inner_impl_scope.AddParent(&impl_scope);
+      CARBON_RETURN_IF_ERROR(TypeCheckPattern(&for_stmt.variable_declaration(),
+                                              std::nullopt, inner_impl_scope,
+                                              ValueCategory::Let));
```

Comment:
If we type-check the range first, we can pass in an expected type here, which should allow us to support things like `for (x: auto in arr)`.

Reply:
thanks, solved!